storage/innobase/handler/ha_innodb.cc

```diff
@@ -16114,6 +16134,13 @@ ha_innobase::external_lock(
 		DBUG_RETURN(0);
 	} else {
 		DEBUG_SYNC_C("ha_innobase_end_statement");
+		export_vars.rows.mutex.wr_lock();
```

Comment:
You may have problems with the `lookup_handler` stats by putting this into `ha_external_unlock` after @vuvova's 3bb5c6b. Try this test:
```
create table t(id int, s date, e date, period for p(s,e), primary key(id, p without overlaps));

insert t values(1, '2020-01-01', '2021-01-01');
update t for portion of p from '2020-02-02' to '2020-03-03' set id = 2;
--source suite/rpl/include/rpl_innodb_rows_counters.inc
flush table t;
--source suite/rpl/include/rpl_innodb_rows_counters.inc
```

Reply:
If I'm right and you don't want to wait for a fix I could suggest you to store the counters only in main handler for a table:
```
ha_innobase *main_handler= static_cast<ha_innobase*>table->file;
main_handler->row.read++;
```
etc.

Reply:
On the second thought, let's make a correct behavior right here.
Add these two lines in the end of `ha_external_lock`:
```c++
       mysql_audit_external_lock(thd, table_share, lock_type);
   }
 
+  if (lookup_handler != this)
+      lookup_handler->ha_external_unlock(table->in_use);
+
   if (MYSQL_HANDLER_RDLOCK_DONE_ENABLED() ||
       MYSQL_HANDLER_WRLOCK_DONE_ENABLED() ||
       MYSQL_HANDLER_UNLOCK_DONE_ENABLED())

```

Reply:
I tried your suggestion. I see no impact on the counters: they are 6,2,1,4 (master) and 5,2,1,4 (replica) with or without applying the code change. Can you check if those numbers make any sense? Also, with your suggested code change, I am observing an assertion failure:
```
#10 0x00007f66990533a2 in __assert_fail (assertion=0x563f66529230 "!n_mysql_tables_in_use", file=0x563f6663b778 "/mariadb/10.11/storage/innobase/trx/trx0trx.cc", line=377, function=0x563f66529247 "void trx_t::free()") at ./assert/assert.c:101
#11 0x0000563f675d9088 in trx_t::free (this=this@entry=0x7f668f3feb80) at /mariadb/10.11/storage/innobase/trx/trx0trx.cc:377
#12 0x0000563f67340e12 in innobase_close_connection (hton=<optimized out>, thd=0x7f6654000d58) at /mariadb/10.11/storage/innobase/handler/ha_innodb.cc:4951
#13 0x0000563f66e1e972 in ha_close_connection (thd=thd@entry=0x7f6654000d58) at /mariadb/10.11/sql/handler.cc:914
#14 0x0000563f66f84b88 in THD::free_connection (this=this@entry=0x7f6654000d58) at /mariadb/10.11/sql/sql_class.cc:1616
#15 0x0000563f66d8aa11 in unlink_thd (thd=0x7f6654000d58) at /mariadb/10.11/sql/mysqld.cc:2851
```
My entire patch is as follows:
```diff
diff --git a/mysql-test/suite/rpl/t/rpl_innodb_bug68220.test b/mysql-test/suite/rpl/t/rpl_innodb_bug68220.test
index 967ac2cf164..567cf4ad106 100644
--- a/mysql-test/suite/rpl/t/rpl_innodb_bug68220.test
+++ b/mysql-test/suite/rpl/t/rpl_innodb_bug68220.test
@@ -22,10 +22,10 @@ select variable_value into @slave_rows_inserted from information_schema.global_s
 --connection master
 CREATE DATABASE testdb;
 USE testdb;
-CREATE TABLE testdb.t1 (i int NOT NULL PRIMARY KEY) ENGINE=InnoDB;
+CREATE TABLE t1 (i int NOT NULL PRIMARY KEY) ENGINE=InnoDB;
 
 # insert a row and show counters on master and slave
-INSERT INTO testdb.t1 VALUES (1);
+INSERT INTO t1 VALUES (1);
 --source suite/rpl/include/rpl_innodb_rows_counters.inc
 
 # update the row and show counters on master and slave
@@ -37,8 +37,16 @@ UPDATE t1 SET i=2 WHERE i=1;
 DELETE FROM t1 WHERE i=2;
 --source suite/rpl/include/rpl_innodb_rows_counters.inc
 
+CREATE TABLE t
+(id INT, s DATE, e DATE, PERIOD FOR p(s,e),
+ PRIMARY KEY(id, p WITHOUT OVERLAPS)) ENGINE=InnoDB;
+INSERT INTO t VALUES(1, '2020-01-01', '2021-01-01');
+UPDATE t FOR PORTION OF p FROM '2020-02-02' TO '2020-03-03' SET id = 2;
+--source suite/rpl/include/rpl_innodb_rows_counters.inc
+FLUSH TABLE t;
+--source suite/rpl/include/rpl_innodb_rows_counters.inc
+
 # clean the test
-DROP TABLE t1;
 DROP DATABASE testdb;
 --sync_slave_with_master
 
diff --git a/sql/handler.cc b/sql/handler.cc
index 83ae4117725..785e7e2797d 100644
--- a/sql/handler.cc
+++ b/sql/handler.cc
@@ -7159,6 +7159,9 @@ int handler::ha_external_lock(THD *thd, int lock_type)
       mysql_audit_external_lock(thd, table_share, lock_type);
   }
 
+  if (lookup_handler != this && lock_type == F_UNLCK)
+    lookup_handler->ha_external_unlock(table->in_use);
+
   if (MYSQL_HANDLER_RDLOCK_DONE_ENABLED() ||
       MYSQL_HANDLER_WRLOCK_DONE_ENABLED() ||
       MYSQL_HANDLER_UNLOCK_DONE_ENABLED())
diff --git a/storage/innobase/handler/ha_innodb.cc b/storage/innobase/handler/ha_innodb.cc
index 67ce245f48c..aed122832af 100644
--- a/storage/innobase/handler/ha_innodb.cc
+++ b/storage/innobase/handler/ha_innodb.cc
@@ -16140,7 +16140,7 @@ ha_innobase::external_lock(
 		export_vars.rows.read += rows.read;
 		export_vars.rows.updated += rows.updated;
 		export_vars.rows.mutex.wr_unlock();
-		new (&rows) typeof(rows) {};
+		rows = {};
 	}
 
 	/* MySQL is releasing a table lock */
```

Reply:
I applied the test case patch to 10.6 bdaa6bac05b2c0cfb8a974a798775a88e124e2de and got the same counter differences: 6,2,1,4 on the master and 5,2,1,4 on the replica.

Reply:
I was wrong about the nature of `ha_reset` it's called for every DML in the end.  So there is no problem with Serg's solution.